ScopesAuthorizer refactoring

mehyaa opened the pull request:
Scopes can be a space separated list in a single claim. Include this possibility on allowed scopes check.

Fixes #913, #1066

## Proposed Changes

  - Split user scope values with space character if scope claim is an single claim and its value has space character
  - User scopes must have all allowed scopes

mehyaa commented:
The build is broken, it fails on irrelevant part from this PR. #1436 fixes is I think.

PatrickDelancy commented:
This is still an issue. Adding my :+1: to get this small and valuable PR merged.

raman-m commented:
Please do the following:
- Merge [PR 1](https://github.com/mehyaa/Ocelot/pull/1)
- Rebase feature branch onto develop
- Resolve all merge conflicts

raman-m commented:
Mehmet,
Thanks for resolving of merge conflicts!

Unfortunately the last build is failed: [5 acceptance tests have failed](https://app.circleci.com/pipelines/github/ThreeMammals/Ocelot/774/workflows/2e281419-906b-4369-8697-af39c0b204ff/jobs/1426?invite=true#step-102-353336_123)!

Why is your PR code so unstable?

mehyaa commented:
I haven't found out the reason why the tests failed with a quick look. I couldn't figure out the tests structure. When I have time I'll look into it.

raman-m commented:
@mehyaa 
The [feature branch](/mehyaa/Ocelot/tree/patch-1) has been [rebased](#event-10182292949) onto [ThreeMammals:develop](../tree/develop) successfully!
The build has failed with [5 tests](https://app.circleci.com/pipelines/github/ThreeMammals/Ocelot/798/workflows/72067b8e-72f4-4a35-a3c5-fbbf844ab0cd/jobs/1474?invite=true#step-102-353716_123)!
**Code review will start after fixing of these failed tests.**

mehyaa commented:
@raman-m I've fixed the tests. Failing tests were written for the bug that requires one of allowed scopes. I've changed the claims and allowed scopes on tests so they can test the correct conditions.

For adding new tests to test `ScopesAuthorizer `, the current tests seem pretty sufficient.

mehyaa commented:
@raman-m I've added you as collaborator on my fork, you can fix the diff or guide me to how-to.

mehyaa commented:
Interestingly some irrelevant tests fails irregularly

raman-m commented:
> @mehyaa commented [on Sep 14, 11](#issuecomment-1719286047)

Thanks for fixing of failed tests.

---
>For adding new tests to test ScopesAuthorizer, the current tests seem pretty sufficient.

No, at least one new test should cover claims logic having them multiple in the related config property.
Simultaneously, we should update current tests to be green. Because each test covers specific atomic feature.

Come on! We've changed the logic from single Scope to multiple ones! And it is definitely right time to cover these changes.

I have idea: let's write tests for each linked issue:
- a test for #913 where you can ask @fsmullinslc to help you
- a test for #1066 where you can ask @papugamichal to help you

Sounds good?

raman-m commented:
> @mehyaa commented [on Sep 14, 11:42 AM](#issuecomment-1719290737)

Thanks for adding me as collaborator!
Now your [develop branch](/mehyaa/Ocelot/tree/develop) is up to date with [ThreeMammals:develop](../tree/develop). So, I've performed [Sync fork](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/syncing-a-fork) operation.
**Done!**
You can start rebasing of current branches onto (creation of new ones from) your **develop** branch.

raman-m commented:
> @mehyaa commented [on Sep 15](#issuecomment-1720779671)

Don't worry! This is unstable scenario: [Ocelot.AcceptanceTests.ConfigurationReloadTests.should_reload_config_on_change](https://app.circleci.com/pipelines/github/ThreeMammals/Ocelot/835/workflows/500d10a4-ae42-4012-9356-c753bd38dc55/jobs/1548?invite=true#step-102-307868_40)
The next run fixes the build usually.
Truly speaking, I am tired of this test too. I will create bug issue soon for this test.

raman-m commented:
@mehyaa 
Why not to continue working? Firstly resolving all merge conflicts...

raman-m commented:
The branch has been [rebased](https://github.com/ThreeMammals/Ocelot/pull/1478#event-12493693341) onto [release/24.0](https://github.com/ThreeMammals/Ocelot/tree/release/24.0) with top commit https://github.com/ThreeMammals/Ocelot/commit/59b63eab9c2cd2ab33c5752bbc3bd00a079cd74c !

raman-m commented:
@mehyaa Mehmet,
If you don't care about your PR anymore, let me to assign to myself. And I'll deliver feature with my vision on design.

### Development is required❕

raman-m commented:
Failed acceptance test: `Ocelot.AcceptanceTests.AuthorizationTests.should_return_response_200_using_identity_server_with_allowed_scope`

alex0360 commented on src/Ocelot/Authorization/ScopesAuthorizer.cs:
```diff
@@ -21,21 +16,32 @@ public Response<bool> Authorize(ClaimsPrincipal claimsPrincipal, List<string> ro
-            var matchesScopes = routeAllowedScopes.Intersect(userScopes);
-
-            if (!matchesScopes.Any())
+            if (routeAllowedScopes.Any(s => !userScopes.Contains(s)))
```
Denial is more costly:

To make scope checking more efficient and less costly, a good practice is to invert the condition.

You can also use a method that minimizes the number of iterations required in the collections, making sure that all required scopes are present in the user scopes.

```c#
if (routeAllowedScopes.TrueForAll(userScopes.Contains))
{
    return new OkResponse<bool>(true);
}

return new ErrorResponse<bool>(
    new ScopeNotAuthorizedError($"User scopes: '{string.Join(",", userScopes)}' do not have all allowed scopes: '{string.Join(",", routeAllowedScopes)}'"));
```

raman-m replied:
Right! Thanks! 
It makes sense to invert the condition.

raman-m commented:
## Predecessor
- #1431 will be merged first

raman-m commented on src/Ocelot/Authorization/ScopesAuthorizer.cs:
```diff
@@ -28,14 +28,22 @@ public Response<bool> Authorize(ClaimsPrincipal claimsPrincipal, List<string> ro
+            }
 
-            if (!matchesScopes.Any())
+            if (routeAllowedScopes.Except(userScopes).Any())
```
This logic inversion is feasible, yet it appears redundant. It seems to be a minor refactoring aimed at reducing the number of lines in the code.
**Finally, it is useless change!**

raman-m commented on src/Ocelot/Authorization/ScopesAuthorizer.cs:
```diff
@@ -28,14 +28,22 @@ public Response<bool> Authorize(ClaimsPrincipal claimsPrincipal, List<string> ro
+            IList<string> userScopes = values.Data;
 
-            var matchesScopes = routeAllowedScopes.Intersect(userScopes);
+            if (userScopes.Count == 1)
```
The `if`-block ought to be relocated to `IClaimsParser` to preserve the existing logic intact. Consequently, you must inject your specialized `IClaimsParser` service to generate the precise list of claims.

raman-m reviewed (changes requested):
Reviewing the currently open PR #1431 would be beneficial, as it involves the development of similar changes, but with a more in-depth analysis of the problem within the context of existing auth-provider functionality.

raman-m commented:
@mehyaa commented [on Sep 14, 2023](https://github.com/ThreeMammals/Ocelot/pull/1478#issuecomment-1719290737)
> I've added you as collaborator on my fork, you can fix the diff or guide me to how-to.

Okay, I just visited your repository and realized I don't have permission to sync the fork because you haven't added me as a collaborator yet.
![image](https://github.com/user-attachments/assets/1ba5659e-45a1-4ee6-b4ac-5eb443ecbe68)
Why are you lying to me? This is ridiculous and unacceptable!

mehyaa commented:
> @mehyaa commented [on Sep 14, 2023](https://github.com/ThreeMammals/Ocelot/pull/1478#issuecomment-1719290737)
> 
> > I've added you as collaborator on my fork, you can fix the diff or guide me to how-to.
> 
> Okay, I just visited your repository and realized I don't have permission to sync the fork because you haven't added me as a collaborator yet.
...
Why are you lying to me? This is ridiculous and unacceptable!

Easy pal, I've added you as collaborator [on Sep 14, 2023](https://github.com/ThreeMammals/Ocelot/pull/1478#issuecomment-1719290737) and you have performed [Sync fork](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/syncing-a-fork) operation [on Sep 15, 2023](https://github.com/ThreeMammals/Ocelot/pull/1478#issuecomment-1721612051). I've not perform any operation on that fork afterwards so I have no idea why you are not a collaborator now. I've added you as collaborator again now but your attitude is not elegant, I have no grudge or what so ever towards you so why should I lie for anything, you can ask again kindly for this or anything within my reach. Later.

raman-m commented:
Thank you for adding me as a collaborator!
Are you planning to address the [code review issues](https://github.com/ThreeMammals/Ocelot/pull/1478#pullrequestreview-2414999052) further to resolve them?

raman-m commented:
> @mehyaa requested a review from @raman-m [on November 19](https://github.com/ThreeMammals/Ocelot/pull/1478#event-21051946911)

You are not following our [development process](https://ocelot.readthedocs.io/en/latest/building/devprocess.html)! The failed [Coveralls build](https://coveralls.io/builds/76580937) stating "no unit testing" confirms this. Following the dev process means writing unit tests for the related issues to cover all new code lines, and then adding acceptance tests to truly verify that the feature works as intended.

So, Mehmet,
I'll review this PR as soon as both unit and acceptance tests are completed.

raman-m commented:
> TO: @mehyaa 

Mehmet,  
Why did you include tests that are failing?
<img width="630" height="617" alt="image" src="https://github.com/user-attachments/assets/96e86624-8ba4-494f-957f-234f8810e52f" />
You wrote the tests with the help of AI but didn’t review them manually in the IDE. The tests failed, and it might surprise you that this is also a violation of the development process, as all AI-generated code must be double-checked, which you didn’t do.

What are you doing in our repo? If this is meant as a contribution to the product, I would prefer not to have it, as it feels like a waste of time.
The next step is to fix the tests, as they seem incomplete or the feature might not be functioning as expected.

## 🟥 Red card penalty❗
You have received a red card for using AI tools. It appears that you and/or the engineers at your company, where you serve as CTO or Lead Engineer, are utilizing AI tools for JavaScript website development. AI vibe-coding approach does not work well with the Ocelot .NET product, as every small detail is crucial.

**P.S. A red card means that if you continue using AI tools to push commits or open new PRs with a help of AI in our repo, your account will be banned. This is the official notice!**

> CC: @ggnaegi @RaynaldM

raman-m commented on src/Ocelot/Authorization/ScopesAuthorizer.cs:
```diff
@@ -28,7 +29,17 @@ public Response<bool> Authorize(ClaimsPrincipal claimsPrincipal, List<string> ro
+        {
+            var scope = userScopes[0];
+
+            if (scope.Contains(' '))
```
1. Both `if`-statements can be merged into a single statement, as having one inside another doesn't look great.
2. Another issue is the hardcoded whitespace character `' '`, which should be defined as a public constant at the class level.

ggnaegi replied:
I don't think the first if statement is needed... please have a look in the ticket feed.

mehyaa replied:
I prefer readability but I've changed as you suggested.

The constant was right, I too do not like magic values.

raman-m commented on test/Ocelot.AcceptanceTests/Authorization/AuthorizationTests.cs:
```diff
@@ -271,6 +271,116 @@ public async Task Should_return_200_OK_with_global_allowed_scopes()
+        GivenThereIsAConfiguration(configuration);
+
+        await GivenThereIsExternalJwtSigningService(route.AuthenticationOptions.AllowedScopes.ToArray());
+        GivenThereIsAServiceRunningOn(port, HttpStatusCode.OK, "Hello from space-separated test");
```
Custom body checking is optional, as is specifying an explicit `HttpStatusCode.OK` status. You can make them implicit by using the response body as the test name:
https://github.com/ThreeMammals/Ocelot/blob/35dbc9d633dc49fe8b5a1f6589ff9e2ec72bf07c/test/Ocelot.AcceptanceTests/Authorization/AuthorizationTests.cs#L265
and
https://github.com/ThreeMammals/Ocelot/blob/35dbc9d633dc49fe8b5a1f6589ff9e2ec72bf07c/test/Ocelot.AcceptanceTests/Authorization/AuthorizationTests.cs#L271

mehyaa commented:
> > TO: @mehyaa
> 
> Mehmet, Why did you include tests that are failing? <img alt="image" width="630" height="617" src="https://private-user-images.githubusercontent.com/10501504/523266763-96e86624-8ba4-494f-957f-234f8810e52f.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjUwMjMwNDMsIm5iZiI6MTc2NTAyMjc0MywicGF0aCI6Ii8xMDUwMTUwNC81MjMyNjY3NjMtOTZlODY2MjQtOGJhNC00OTRmLTk1N2YtMjM0Zjg4MTBlNTJmLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTEyMDYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUxMjA2VDEyMDU0M1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWM4N2M3NTJlNTVmMmQ3YTkxMTllYWIwMmY0NzZjZWJmZjhhMTUyYzQxOWVlMWViODFmOGM5NGZlMDVkZjVmOGEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.caAArK2eIHDfDLpCaS0oELHvuy3rnsLDJT0AsuT51mg"> You wrote the tests with the help of AI but didn’t review them manually in the IDE. The tests failed, and it might surprise you that this is also a violation of the development process, as all AI-generated code must be double-checked, which you didn’t do.
> 
> What are you doing in our repo? If this is meant as a contribution to the product, I would prefer not to have it, as it feels like a waste of time. The next step is to fix the tests, as they seem incomplete or the feature might not be functioning as expected.
> 
> ## 🟥 Red card penalty❗
> You have received a red card for using AI tools. It appears that you and/or the engineers at your company, where you serve as CTO or Lead Engineer, are utilizing AI tools for JavaScript website development. AI vibe-coding approach does not work well with the Ocelot .NET product, as every small detail is crucial.
> 
> **P.S. A red card means that if you continue using AI tools to push commits or open new PRs with a help of AI in our repo, your account will be banned. This is the official notice!**
> 
> > CC: @ggnaegi @RaynaldM

I give up.

I just wanted to fix a bug in the product that I've already fixed in my own implementation. I don't strictly need the fix to be merged, thanks to the flexibility of the project (being able to register your own implementation).

I added the tests manually at first, but the codebase changed dramatically over time. My merge attempts failed, and you had to take over the branch, after which the PR went stale. When I recently returned to finish this work, I implemented the tests with Copilot, verified them, and pushed. However, in the meantime, the acceptance test structure was refactored again, making the PR unmergeable.

I tried fixing the tests manually to match the existing ones so I could build, but I couldn't make them pass (the token endpoint returns 404). I suspected my environment, so I wanted to see the results in the actions; that is why I pushed it.

I know I may not have followed your exact workflow, so you have every right to criticize. I have no intention of bloating or breaking your product; I like and use Ocelot. My sole intention was to fix a minor bug. If my work isn't suitable, I'm stepping away from it plain and simple, with no hard feelings.

Sorry for taking up your time. Thank you and all the contributors for this great product.

mehyaa commented:
> Why userScopes.Count == 1 ?

That will be a error if a provider supplied `scope` claim as an array that some of the values are space-separated.

raman-m commented:
Oops, I've changed my mind. I forgot that PR is now part of the [.NET 10](https://github.com/ThreeMammals/Ocelot/milestone/13) milestone.

raman-m commented:
@ggnaegi Why am I not a CTO or super IT manager? Maybe because I'm the one writing the code? 😉
My LinkedIn profile includes a legend:
> A Boy Who Codes! Since 2003

Doesn't this sentence violate human rights in Switzerland, and maybe in EU? 
To be fair, I've been coding since school, when my first program was a BASIC script that played a sound and displayed text on the screen of a legendary monochrome PC display 💾 

Anyway, let me focus on the current hot release of v24.1...

mehyaa commented:
> > > Why userScopes.Count == 1 ?
> > 
> > That will be a error if a provider supplied `scope` claim as an array that some of the values are space-separated.
> 
> @mehyaa could you just comment this in code please. It's fine for me and i will approve your PR. 
> 

Added a comment to the code.

raman-m commented on src/Ocelot/Authorization/ScopesAuthorizer.cs:
```diff
@@ -28,7 +31,14 @@ public Response<bool> Authorize(ClaimsPrincipal claimsPrincipal, List<string> ro
+
+        // There should not be more than one scope claim that has space-separated value by design
+        // Some providers use array value some space-separated value but not both
+        if (userScopes.Count == 1 && userScopes[0].Contains(' '))
```
```suggestion
        if (userScopes.Count == 1 && userScopes[0].Contains(SpaceChar))
```

mehyaa replied:
My case was [auth0](https://auth0.com). The [RFC ](https://www.rfc-editor.org/rfc/rfc8693#section-4.2) states the scope claim is a space-separated value but sadly implementations varies as I can see.

raman-m commented on src/Ocelot/Authorization/ScopesAuthorizer.cs:
```diff
@@ -28,7 +31,14 @@ public Response<bool> Authorize(ClaimsPrincipal claimsPrincipal, List<string> ro
         }
 
-        var userScopes = values.Data;
+        IList<string> userScopes = values.Data;
```
This change isn't necessary; using `var` is fine.
The problem is that you are casting to a more abstract type, which can slightly affect performance.
What was your reason for casting the `List<string>` object?

mehyaa replied:
Arrays and Lists share the IList implementation so I can use the same pointer for both the list from the initial value and then if space-sapareted array from the `string.Split`. There is no boxing there. Later the instance will be used by the LINQ anyway. If I use `var` then I need to call `ToList` on the `string.Split` value which will be (slightly) worse performance.

raman-m commented on test/Ocelot.AcceptanceTests/Authentication/AuthenticationSteps.cs:
```diff
@@ -101,9 +101,9 @@ public void WithJwtBearerAuthentication(IServiceCollection services, bool addOce
         app.MapPost("/token", (AuthenticationTokenRequest model) =>
         {
-            if (!apiScopes.Contains(model.Scope))
+            if (!apiScopes.Intersect(model.Scopes.Split(' ')).Any())
```
The code looks fine.
Since this isn't part of the core library, heavy use of LINQ to Objects isn't a concern.
FYI, I might change this logic in future PRs, but the splitting will remain.

mehyaa commented on test/Ocelot.AcceptanceTests/Authorization/AuthorizationTests.cs:
```diff
@@ -271,6 +271,110 @@ public async Task Should_return_200_OK_with_global_allowed_scopes()
+    [Fact]
+    [Trait("PR", "1478")]
+    [Trait("Feat", "913")]
+    [Trait("Feature", "ScopesAuthorizer: Space-separated scopes")]
```
Removed the redundant traits.

raman-m reviewed (approved):
## Ready for delivery ✅ 
- Code review ✔️ ✔️ 
- Unit testing ✔️ ✔️ 
- Acceptance testing ✔️ 
- Updated docs ✔️
